.github/workflows/test.yml

```diff
@@ -37,8 +37,244 @@ jobs:
       matrix:
         os: [ubuntu-latest]
         ghc: [9.0.2, 8.10.7, 8.8.4, 8.6.5]
-        z3: [4_8_8, 4_8_9, 4_8_10, 4_8_11, 4_8_12, 4_8_13, 4_8_14]
-        allow-failure: [false]
+        z3: [ 4_8_14 ]
+        yices: [ 2_6 ]
+        stp: [ 2_3_3 ]
+        cvc4: [ 1_8 ]
+        boolector: [ 3_2_2 ]
+        abc: [ 2021_12_30 ]
+        allow-failure: [ false ]
+        # Specify multiple older solver versions for the latest builds
+        # to check backward compatibility.  These older solvers are
+        # not simply added to the general matrix above to avoid
+        # creating excessive numbers of repetitive jobs.
+        include:
```

Comment:
Note for the reviewers: github syntax here is regrettably limited.  Above, we specify the matrix of versions we are interested in.  By default, github will test all combinations in that matrix, but that would create a *very* large target space.  To address this, I've limited the specifications in the above matrix to the current versions of solvers, and then used the `include` section here to include additional target configurations I'd like to test.  I use the `include` to run tests on older solver versions (here I'm testing older solver versions for both GHC 8.10.7 and 9.0.2 since I consider 9.0.2 to still be probationary).

Unfortunately, the syntax for a github `include` requires the explicit specification of all parameters, even if they are going to be the same as in the base matrix above.  Thus, there's a lot of repetition below and some comments describing which field is actually being modified.  It's verbose.  I hate it, but it's the best way I know of.

There's also an `exclude` syntax, which lets you specify which points of the original matrix you do not want tested.  It has the same limitations as the `include`: all parameters must be specified even if they haven't changed.  I believe this route would have yielded even more statements than the minimal matrix + include method I chose.

It's verbose.  I hate it.  I don't

Reply:
I hate this too, and I especially hate the idea of having to maintain this. I'm wondering if we should specify the `include` matrix [as JSON](https://docs.github.com/en/actions/learn-github-actions/expressions#fromjson) instead. That is, have a job that runs before the `linux` job that generates a JSON representation of the `include` matrix as output. Then, the `linux` job can specify the `include` matrix like so:

```yaml
include: ${{fromJson(needs.matrix_prep_job.outputs.matrix)}}
```

See [here](https://stackoverflow.com/a/65434401) for an example of how this can be done. This would allow us to consolidate the logic for specifying the different combinations of solvers into a script of some sort, which seems like a much more maintainable solution in the long run.

Reply:
Oh interesting... I wasn't aware of that fromJSON capability. I'll do some noodling on this...

Reply:
The matrix is now generated more maintainably.